internal/server/storage/storage.go

```diff
@@ -0,0 +1,159 @@
+package storage
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+
+	"github.com/npavlov/go-metrics-service/internal/domain"
+	"github.com/npavlov/go-metrics-service/internal/model"
+	"github.com/npavlov/go-metrics-service/internal/server/config"
+	"github.com/npavlov/go-metrics-service/internal/server/snapshot"
+)
+
+const (
+	errNoValue = "no value provided"
+)
+
+type Repository interface {
+	Get(name domain.MetricName) (*model.Metric, bool)
+	Create(metric *model.Metric) error
+	GetAll() map[domain.MetricName]model.Metric
+	Update(metric *model.Metric) error
+	WithBackup(ctx context.Context, cfg *config.Config) *MemStorage
+}
+
+type MemStorage struct {
+	mu       *sync.RWMutex
+	metrics  map[domain.MetricName]model.Metric
+	cfg      *config.Config
+	l        *zerolog.Logger
+	snapshot snapshot.Snapshot
+}
+
+// NewMemStorage - constructor for MemStorage.
+func NewMemStorage(l *zerolog.Logger) *MemStorage {
+	ms := &MemStorage{
+		metrics:  make(map[domain.MetricName]model.Metric),
+		mu:       &sync.RWMutex{},
+		l:        l,
+		cfg:      nil,
+		snapshot: nil,
+	}
+
+	return ms
+}
+
+func (ms *MemStorage) WithBackup(ctx context.Context, cfg *config.Config) *MemStorage {
+	memSnapshot := snapshot.NewMemSnapshot(cfg.File, ms.l)
+	ms.snapshot = memSnapshot
+	ms.cfg = cfg
+
+	if cfg.RestoreStorage {
+		metrics, err := memSnapshot.Restore()
+		if err != nil {
+			ms.l.Error().Err(err).Msg("failed to restore metrics")
+		}
+		if err == nil {
+			ms.metrics = metrics
+			ms.l.Info().Msg("Metrics restored successfully")
+		}
+	}
+
+	ms.startBackup(ctx)
+
+	return ms
+}
+
+func (ms *MemStorage) startBackup(ctx context.Context) {
+	if ms.cfg.StoreInterval > 0 {
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					ms.l.Info().Msg("Stopping storage backup")
+
+					return
+				default:
+					ms.mu.RLock()
+					err := ms.snapshot.Save(ms.metrics)
+					ms.mu.RUnlock()
+					if err != nil {
+						ms.l.Error().Err(err).Msg("Error saving file")
+						panic(err)
+					}
+					time.Sleep(time.Duration(ms.cfg.StoreInterval) * time.Second)
```

Comment:
[CHANGE REQUEST] It should be `time.Duration`

Reply:
startBackup is not public method 

ms.cfg.StoreInterval type is time.duration